botocore/client.py

```diff
@@ -325,6 +249,155 @@ def _api_call(self, *args, **kwargs):
+            'endpoint_url': endpoint_url,
+            'signature_version': signature_version,
+            'metadata': metadata
+        }
+
+    def _make_url(self, hostname, is_secure):
+        endpoint_url = 'https://' if is_secure else 'http://'
+        return endpoint_url + hostname
+
+    def _resolve_signing_name(self, service_name, resolved):
+        if 'credentialScope' in resolved \
+                and 'service' in resolved['credentialScope']:
+            return resolved['credentialScope']['service']
+        return service_name
+
+    def _pick_region_values(self, resolved, region_name, endpoint_url):
+        signing_region = region_name
+        if endpoint_url is None:
+            # Do not use the region name or signing name from the resolved
+            # endpoint if the user explicitly provides an endpoint_url. This
+            # would happen if we resolve to an endpoint where the service has
+            # a "defaults" section that overrides all endpoint with a single
+            # hostname and credentialScope. This has been the case historically
+            # for how STS has worked. The only way to resolve an STS endpoint
+            # was to provide a region_name and an endpoint_url. In that case,
+            # we would still resolve an endpoint, but we would not use the
+            # resolved endpointName or signingRegion because we want to allow
+            # custom endpoints.
+            region_name = resolved['endpointName']
+            signing_region = region_name
+            if 'credentialScope' in resolved \
+                    and 'region' in resolved['credentialScope']:
+                signing_region = resolved['credentialScope']['region']
+        return region_name, signing_region
+
+    def _resolve_signature_version(self, service_name, resolved):
+        # Client config overrides everything.
+        client = self.client_config
+        if client and client.signature_version is not None:
+            return client.signature_version
+        # Scoped config overrides picking from the endpoint metadata.
+        scoped = self.scoped_config
+        if scoped is not None:
+            service_config = scoped.get(service_name)
+            if service_config is not None and isinstance(service_config, dict):
+                version = service_config.get('signature_version')
+                if version:
+                    logger.debug(
+                        "Switching signature version for service %s "
+                        "to version %s based on config file override.",
+                        service_name, version)
+                    return version
+        # Pick a signature version from the endpoint metadata if present.
+        if 'signatureVersions' in resolved:
+            for preference in self.SIGNATURE_PRECEDENCE:
+                if preference in resolved['signatureVersions']:
+                    return preference
+        raise UnknownSignatureVersionError(
```

Comment:
This should probably not fail here unless we actually don't know the signature version.  The AUTH_TYPE_MAP map in auth.py has the definitive source of which signature versions we support.  There's a couple of other signature versions in with AUTH_TYPE_MAP that we support that will now fail, e.g `v3`.  While nothing uses that signature version anymore, it still seems like we should only claim an `UnknownSignatureVersionError` if it's truly an unknown signature version.